image-registry/automate-pruning.md

```diff
@@ -0,0 +1,300 @@
+Specifically, this includes the following:
+
+1. A `ServiceAccount` dedicated to the image pruner.
+2. A `ClusterRoleBinding` granting the `system:image-pruner` role to the pruner
+service account.
+3. A `CronJob` which runs the image pruner with the following tunable parameters:
+   1. Schedule for `CronJob`
+   2. `--keep-tag-revision` command line argument for the pruner
+   3. `--keep-younger-than` command line argument for the pruner
+   4. Resources for the `CronJob` pod template
+
+Installation and configuration of the pruner will be accomplished via a separate CRD.
+These components should be created/reconciled if the pruning CRD is created or updated.
+When and how this CRD is created will vary on the graduation phase of this feature.
+
+When the pruning CRD is deleted, the pruning `CronJob` and its related components
+should also be deleted.
+
+The operator's behavior with respect to managing the pruner is independent of
+the `ManagementState` specified on the image registry operator's `ClusterOperator` object.
+
+### Risks and Mitigations
+
+#### Resource consumption
+
+Risk: Image pruning requires siginficant memory to compute the blob layer graph
+needed to identify "prunable" layers.
+
+Mitigation: API will allow cluster admins to adjust the resources allocated to
+the `CronJob`.
+
+Risk: Image pruning requires too much memory to be scheduled to a node
+Risk: Image pruning exhausts resources used for application workloads
+
+Mitigation: API will allow admins to use standard tools to let pruning be run
+on desired machine pools (`affinity`, `tolerations`, `nodeSelector`)
+
+##### Concurrent pruning jobs
+
+Risk: Pruning jobs cannot run concurrently.
+Risk: Pruning job durations can exceed the interval set by the `CronJob`
+schedule.
+
+Mitigation: The `CronJob` for pruning must be configured with
+`concurrencyPolicy: Forbid`
+
+#### Removal of past pruning jobs
+
+Risk: Runs of the pruning job create another set of objects in etcd that need
+to be pruned.
+
+Mitigation: `CronJob` specifies a history limit for successful and failed jobs.
+This will be a tunable parameter via the API.
+
+#### Pruning of in-use images
+
+Risk: Pruner is not aware of a particular image, resulting in it being pruned
+prior to use.
+
+Mitigation: None at present - this is an inherent problem within the current
+pruning process.
+
+## Design Details
+
+### API
+
+A separate CRD will be created with the following `spec` and `status` fields:
+
+```yaml
+spec:
+  schedule: "*/1 * * * *"
+  keepTagRevisions: 3
+  keepYoungerThan: 60m
+  resources: {}
+  affinity: {}
+  nodeSelector: {}
+  tolerations: {}
+  history:
+    successfulJobsHistoryLimit: 3
+    failedJobsHistoryLimit: 1
+status:
+  observedGeneration: 2
+  conditions:
+  - type: Scheduled
+    status: "True"
+    lastTransitionTime: 2019-10-09T03:13:45
+    reason: Scheduled
+    message: "Image pruning has been scheduled."
+  - type: Completed
```

Comment:
`InProgress` seems more intuitive to me, but I can defer if others do not agree

Reply:
Is `Completed` information about the latest successful run? I'd prefer not to have a condition that is by design going to periodically change its state.

Reply:
> InProgress seems more intuitive to me, but I can defer if others do not agree

@gabemontero `Completed` comes from the Job API [1] - if we're bubbling up info from other API objects, I'd rather stay consistent.

> Is Completed information about the latest successful run? I'd prefer not to have a condition that is by design going to periodically change its state.

@dmage it is about the last run (success or failed). My intent was to make it easier to access this information, as `CronJob` does not directly indicate if the last run succeeded or not.

[1] https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#jobcondition-v1-batch

Reply:
ok works for me thanks

Reply:
@dmage would it make more sense if we exposed the net status of the pruning jobs as Prometheus metrics? Ex - a gauge with the number of active pruning jobs (should be 1 or 0), and a counter of the completed pruning jobs by result (result="succeeded"|"failed"). The latter seems necessary if we want to alert on failure.

Reply:
@adambkaplan we definitely will want to have metrics and alerts. I can imagine two alerts:

1. the latest pruning job has finished unsuccessfully
2. the latest pruning job works for more than 1 day

I'm not sure about counters though, it might be hard to count jobs when they are getting pruned.

Reply:
Prometheus counters are interesting - they work as an "increment only" indicator for things like HTTP requests [1]. On paper we could watch the pruning jobs in the namespace and update metrics.

In any case I don't think we should sweat the details on the metrics too much - during development we can decide which kind of metric is appropriate to drive a desired alert. I'll update the telemetry section with the alerts we want.

[1] https://prometheus.io/docs/concepts/metric_types/#counter

Reply:
Regarding conditions, I'm ok with having conditions about the latest success/failure, but the name `Completed` is ambiguous for our resource, it works only for Jobs (they can have either `Completed` or `Failed`). Upstream decided not to propagate information from Jobs to CronJobs, so we don't have an easy answer for this. I'd rather have a single condition whether the latest run is successful or not.

Reply:
Restored the `Failed` condition so we can provide visibility on pruning job failures.